### PR TITLE
Add support for type modifiers on extension types

### PR DESCRIPTION
**Pull request by Maxxen**
This PR introduces support for adding "type modifiers" on custom types created through extensions. 

### Explanation

"Type modifiers" are basically an optional ordered sequence of `duckdb::Value`s attached to a `LogicalType` which can be used to store some type-level metadata or properties to enable a limited form of parametric polymorphism. We call types accepting type modifiers "parameterized types'

A good example of a type with "modifiers" that already exists in DuckDB today is the `DECIMAL(WIDTH, SCALE)` type, where the `WIDTH` and `SCALE` parts of the type allows users to create many different kinds of decimals by "modifying" the type parameters. DuckDB itself can then provide functions that are generic over any particular decimal type instance, instead of having to create a new concrete type and function overload for every possible combination of decimal.

While this PR leaves the existing types (such as the DECIMAL) unchanged, it generalizes this functionality so that it is now possible for extensions to create similarly parameterized types. Here's an example from the test extension:

```sql
-- The `loadable_extension_demo` extension now provides a `MINMAX(INT, INT)` type that is 
-- an alias  over int32 , but also takes two type parameters that can be inspected during 
-- casting/function execution and used to enforce that the value is within a min-max range.

SELECT 5::MINMAX(0, 10)
----
5

SELECT 15::MINMAX(0, 10);
----
Conversion Error: Value 15 is outside of range [0,10]

SELECT 10::MINMAX(15, 10);
----
Binder Error: MINMAX type min value must be less than max value

SELECT 10::MINMAX(1337);
----
Binder Error: MINMAX type must have two modifiers

CREATE TABLE minmax_table (i MINMAX(0, 100));

INSERT INTO minmax_table VALUES (10), (250)
----
Conversion Error: Value 250 is outside of range [0,100]

-- This is a function that only inspects the modifiers
SELECT minmax_range(3::MINMAX(2, 7));
----
 5 

```

### Type binding

If an extension registers a custom type with type modifiers already set, the types of these modifiers will be used to resolve the type modifiers supplied by a user when instantiating the custom type through SQL. E.g. if you register a type `{ type: INT32, alias: 'MY_TYPE', modifiers: [5::INT32, 10::INT64] }` and a user then issues `CREATE TABLE t1 (val MY_TYPE(<m1>, <m2>))`, the binder will try to resolve `<m1>` to a INT32 and `<m2>` to a INT64, casting if necessary as that is the types of the "original" modifiers supplied when registering the custom type `MY_TYPE`. 

The original type modifiers also function as "default" values, should the user not supply the full set of expected type modifiers. E.g. `SELECT typeof(MY_TYPE(2)) == 'MY_TYPE(2, 10)`

Besides providing default type modifiers, an extension can also register custom types with a `bind_type_modifiers_function_t(ClientContext &context, LogicalType &type,
                                               const vector<Value> &modifiers)` function pointer to customize the binding logic of type parameters. If this callback is supplied the binder will invoke it instead of trying to infer the valid type modifiers by looking at the defaults in the originally registered type. This gives the extension a chance to verify that the type modifiers are correct/valid. For example, the `MINMAX` type example above uses a custom type modifier bind function to check that at least two modifiers are supplied and that the "min" value is not larger than the "max" value.

### Function binding 

When comparing two parameterized types (of the same type) for equality, the type modifiers are compared positionally in order by type and value. If the number of type modifiers differ, the additional modifiers not present in one of the types are ignored. Additionally, a typed `NULL` value will compare true to anything as long as the types match.
These semantics make it possible to pattern match, generalize and specialize when defining the types of functions.

For example:
```sql
-- all the common modifiers match
TYPE(10) == TYPE(10, 20)
-- all the common modifiers match - vacuously true since one side doesn't have any modifiers
TYPE == TYPE(10)
TYPE == TYPE(10, 20)
TYPE == TYPE(10) == TYPE(10, 20)

-- typed NULL functions as a placeholder, equals anything
TYPE(10) == TYPE(NULL::INT)
TYPE(10) == TYPE(NULL::INT, 20)

-- the common modifiers don't match because the values are different
TYPE(10) != TYPE(20)
-- the common modifiers don't match (because the type doesn't match)
TYPE(10) != TYPE('foo')
-- untyped null doesn't match (because the type doesn't match)
TYPE(10) != TYPE(NULL, 20)
```

A function `foo` defined as `INT foo(param TYPE)` will accept a `TYPE(10)` but a function `bar` defined as `INT BAR(param TYPE(5)` will not bind as the modifiers don't match.

### CODE REVIEW

**Comment by Maxxen** [comment]
@Mytherin CI failures are unrelated

File: src/include/duckdb/common/types.hpp [hunk 1]
```diff
@@ -273,6 +273,8 @@ struct LogicalType {
 		return type_info_;
 	}
 
+	void MakeTypeInfoUnique();
```
Comment (Mytherin) [inline src/include/duckdb/common/types.hpp:276]: Can we add a `DeepCopy()` method instead?

File: src/include/duckdb/parser/parsed_data/create_type_info.hpp [hunk 2]
```diff
@@ -15,16 +15,22 @@
 
 namespace duckdb {
 
+//! The type to bind type modifiers to a type
+typedef void (*bind_type_modifiers_function_t)(ClientContext &context, LogicalType &type,
+                                               const vector<Value> &modifiers);
```
Comment (Mytherin) [inline src/include/duckdb/parser/parsed_data/create_type_info.hpp:20]: `BindTypeModifierInfo`

File: src/parser/transform/helpers/transform_typename.cpp [hunk 3]
```diff
@@ -131,109 +196,77 @@ LogicalType Transformer::TransformTypeName(duckdb_libpgquery::PGTypeName &type_n
...
 				}
 			}
 			break;
 		default:
-			if (modifier_idx > 0) {
+			if (modifiers.count > 0) {
 				throw ParserException("Type %s does not support any modifiers!", LogicalType(base_type).ToString());
 			}
 			result_type = LogicalType(base_type);
-			break;
```
Comment (Mytherin) [inline src/parser/transform/helpers/transform_typename.cpp:236]: nit

File: src/include/duckdb/common/types.hpp [hunk 4]
```diff
@@ -319,6 +321,12 @@ struct LogicalType {
 	DUCKDB_API void SetAlias(string alias);
 	DUCKDB_API bool HasAlias() const;
 	DUCKDB_API string GetAlias() const;
+	DUCKDB_API void SetModifiers(vector<Value> modifiers);
+	DUCKDB_API bool HasModifiers() const;
+	DUCKDB_API vector<Value> GetModifiers() const;
+	// ! Try to get the modifiers without copying, throws an exception if the type has no modifiers
+	DUCKDB_API vector<Value> &GetModifiersUnsafe();
```
Comment (Mytherin) [inline src/include/duckdb/common/types.hpp:328]: `TryGetModifiers` that returns `optional_ptr<vector<Value>>`

File: src/include/duckdb/common/types.hpp [hunk 5]
```diff
@@ -319,6 +321,12 @@ struct LogicalType {
 	DUCKDB_API void SetAlias(string alias);
 	DUCKDB_API bool HasAlias() const;
 	DUCKDB_API string GetAlias() const;
+	DUCKDB_API void SetModifiers(vector<Value> modifiers);
+	DUCKDB_API bool HasModifiers() const;
+	DUCKDB_API vector<Value> GetModifiers() const;
```
Comment (Mytherin) [inline src/include/duckdb/common/types.hpp:326]: `GetModifiersCopy()`

File: src/planner/binder/statement/bind_create.cpp [hunk 6]
```diff
@@ -211,54 +212,115 @@ void Binder::BindLogicalType(ClientContext &context, LogicalType &type, optional
...
+
 		if (catalog) {
 			// The search order is:
 			// 1) In the same schema as the table
 			// 2) In the same catalog
 			// 3) System catalog
-			type = catalog->GetType(context, schema, user_type_name, OnEntryNotFound::RETURN_NULL);
-
-			if (type.id() == LogicalTypeId::INVALID) {
-				type = catalog->GetType(context, INVALID_SCHEMA, user_type_name, OnEntryNotFound::RETURN_NULL);
```
Comment (Mytherin) [inline src/planner/binder/statement/bind_create.cpp:249]: Maybe the whole `GetType` method can be removed?

File: src/include/duckdb/parser/parsed_data/create_type_info.hpp [hunk 7]
```diff
@@ -15,16 +15,22 @@
 
 namespace duckdb {
 
+//! The type to bind type modifiers to a type
+typedef void (*bind_type_modifiers_function_t)(ClientContext &context, LogicalType &type,
```
Comment (Mytherin) [inline src/include/duckdb/parser/parsed_data/create_type_info.hpp:19]: Let's make this take a `const LogicalType &type` and return a `LogicalType` instead

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good - some minor comments

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**